### ISSUE TITLE
Ranch embedded mode cannot be used twice in a release

### ISSUE THREAD

**Issue by pallix** [ninenines/ranch#211]
My team and I have an application which uses `ranch` for two different TCP servers running in the same application. One TCP server is used for external clients and one is for internal usage, they implement different protocols.

The main application has two applications libraries as dependencies. Since we are using `ranch` in embedded mode and building the supervisor tree ourselves, each of the library defines `included_applications: [:ranch]` to avoid starting the ranch app.

This works well except when releasing: distillery does not allow an application to be included `included_application` more than once. See this [discussion](https://groups.google.com/forum/#!topic/elixir-lang-core/5ZJSZQs8OjU) for an explanation.

Would it be possible for ranch to implement an option so to not start the supervisor? or what solution do you propose?

**Comment by essen** [comment]
I'm not sure there's much to do for me, `included_applications` can only include an application once by design. It essentially embeds the whole application instead of just a few processes. Ranch provides `ranch:child_spec` to embed just the listener processes, maybe that's what you'd want to use instead. Of course in that case there's still the matter of the `ranch_server` which is currently unique and can't be tied directly to the embedded listener processes.

**Comment by zoltanmaric** [comment]
Hi @essen, thanks for making and maintaining this library :)

Our setup is the following:
```
                          app
                        /     \
lib1 with embedded ranch       lib2 with embedded ranch
```
Because they use embedded `ranch`, both `lib1` and `lib2` list `ranch` in `included_applications`, which doesn't work in `app`, because two of its dependencies list `ranch` as included.

What we would actually need, and find reasonable, is to freely start `ranch` as a dependency, but configuring it so that `ranch_sup` is not started.

It somehow seems strange that it is possible to use `ranch` in embedded mode using `ranch:child_spec`, but there's no way not to start `ranch_sup` (which, as far as I understand, you start yourself in embedded mode) without a quite complicated additional dependency setup.

**Comment by essen** [comment]
Sounds like what you want is `{load, ranch}` in relx.config. Or in other terms, make your release contain Ranch but not start it.

**Comment by zoltanmaric** [comment]
@essen yes, this is the solution we ended up using. However, I find it quite awkward, and non-transparent.

Would you not agree that, if `ranch` supports embedded mode and even provides helper functions for it, it would make sense to be able to start it without a standalone `ranch_sup`?

**Comment by essen** [comment]
I'm not sure what the point is? It's not going to be easier than the load option.

**Comment by zoltanmaric** [comment]
Sorry for not being clear enough. My idea was to provide a way that [`ranch_app:start/2`](https://github.com/ninenines/ranch/blob/0d9ad0a8445e7ba6e8bf9591257fa4457dda8321/src/ranch_app.erl#L22) be passed a configuration, or load a configuration from the application environment which tells it not to [start `ranch_sup`](https://github.com/ninenines/ranch/blob/0d9ad0a8445e7ba6e8bf9591257fa4457dda8321/src/ranch_app.erl#L24).

That way it would be sufficient to configure how `ranch` is started only in applications that directly depend on it (`lib1` and `lib2` in our case), and applications further up the dependency tree wouldn't have to care about `ranch` at all.

**Comment by essen** [comment]
There's probably some misunderstanding.

First, if you are indeed using using http://erlang.org/doc/design_principles/included_applications.html you should note this in the introduction:

    An application can only be included by one other application.

This means you can't include Ranch twice.

Now while you don't need `ranch_sup`, you do need `ranch_server` and so you will need to start it somewhere. It's not really `ranch_sup` that is the problem, and it's not possible to run Ranch listeners without `ranch_server`.

So all you need to do is start `ranch_server` somewhere (or just let Ranch be started normally) and stop using `included_applications`.

**Comment by pallix** [comment]
> Now while you don't need ranch_sup

I am surprise to read that since the documentation section on embedding ranch explicitly uses `ranch_sup`.

So there is another way of starting ranch for the embedded mode? Could you give an example or update the documentation?

**Comment by essen** [comment]
You can start `ranch_server` directly without going through `ranch_sup`. That doesn't change the problem much though since you can only have one `ranch_server`.

**Comment by essen** [comment]
@juhlig We will want to improve this for Ranch 2.0 as well.

**Comment by essen** [comment]
I don't think we can cut `ranch_server` in half without losing useful functionality (we'll always need an ets table somewhere even with multiple processes). On the other hand it should be possible to proxy exit signals to `ranch_server` via a good use of links.

Let's say you have appA, appB both depending on Ranch. Both appA and appB use embedded listeners. The embedded listeners need to exit when `ranch_server` exits to avoid keeping inconsistent state. This can be accomplished using a supervisor similar to `ranch_sup` except that in place of `ranch_server` you would have a proxy process that links itself to `ranch_server`. When `ranch_server` exits, the proxy process exits and its supervisor will in turn terminate the listener.

We can probably provide this proxy process functionality in Ranch itself. The only difficulty is during restarts because the proxy process might restart before `ranch_server`, so the proxy process will need some intermediary state where it's not linked to `ranch_server`. It can probably stay in this state indefinitely but would need to log information and perhaps expose a metric indicating it's not properly working.

Thoughts?

**Comment by essen** [comment]
This should be fixed with the work in #249. The child spec returned by Ranch now includes a module that monitors `ranch_server` to keep the same behavior as if you were in the Ranch supervision tree. This still requires Ranch to be started however, and that's a limitation we are not going to remove. We will no longer suggest having `ranch_server` or `ranch_sup` started out of the Ranch supervision tree. And, no change there, `included_applications` will not be expected to work for two or more separate applications.

I think this can be closed, I'll merge the mentioned PR in a few.

**Comment by essen** [comment]
Thanks!
